lib/BloqadeSchema/src/transform.jl

```diff
@@ -0,0 +1,389 @@
+# this function is required to calculate difference
+# between waveforms which have different durations.
+# the behavior is to pad the shorter waveform at the
+# end with 0. This is useful when the user passes
+# in a duration that doesn't neatly fit into the
+# time resolution of the device. In which case 
+# this is reqiured to check the error.
+function norm_diff_durations(A::Waveform,B::Waveform)
+    A.duration==B.duration && return norm(A-B)
+
+    if A.duration > B.duration
+        T_diff = A.duration - B.duration
+        B = append(B,constant(;duration=T_diff,value=0))
+        return norm(A-B)
+    else
+        T_diff = B.duration - B.duration
+        A = append(A,constant(;duration=T_diff,value=0))
+        return norm(A-B)
+    end
+end
+
+# simple function, round number to the nearest multiple of `res`
+function set_resolution(x::Real,res::Real)
+    @assert res > 0
+    return round(Int(round(x / res)) * res,sigdigits=14)
+end
+
+# throws error for none-scalar fields
+function  check_global(field,name) 
+    field isa AbstractArray && error("Failed to transform $name to hardware, $name must be global drive.")
+end
+
+# checks argument to make sure it is a waveform or a container of waveforms.
+function check_waveform(field,name)
+    field isa Waveform{F,T} where {F,T<:Real} && return 
+        
+    field isa Vector{<:Number} && error("Failed to transform $name to hardware, $name must be a vector Waveforms.\n If $name contains constant values use `BloqadeWaveforms.constant`.")
+    if field isa Vector 
+        foreach(field) do ele
+            !(ele isa Waveform{F,T} where {F,T<:Real})  && error("Failed to transform $name to hardware, $name must be a Waveform.")
+        end
+
+        duration = field[1].duration
+        foreach(field) do ele
+            duration != ele.duration && error("Failed to transform $name to hardware, all Waveforms must have the same duration.")
+        end
+
+        return
+
+    end
+    (isnothing(field) || field isa Number || BloqadeExpr.is_time_function(field)) && error("Failed to transform $name to hardware, $name must be a Waveform.\n If $name is constant or zero use `BloqadeWaveforms.constant`.")
+
+end
+# warns user if the duration of the waveform will be rounded by the time resolution.
+function warn_duration(time_res,field,name)
+    duration = if field isa Vector
+        field[1].duration
+    else
+        field.duration
+    end
+    if !(set_resolution(duration,time_res) ≈ duration)
+        @info "waveform $name duration will be rounded during hardware transformation."
```

Comment:
@johnzl-777

Reply:
This causes existing tests based on `@test_logs (:debug, ...)` to fail. They'll need to be changed to something like 
```Julia
using Logging
@test_logs (:debug,"waveform test duration will be rounded during hardware transformation.") min_level=Logging.Debug BloqadeSchema.warn_duration(time_res,wf,:test)
```
(modified from: https://docs.julialang.org/en/v1/stdlib/Test/#Test.@test_logs)
It may require we add `Logging` as a dependency for the tests.

Reply:
that's fine, `Logging` is a stdlib

Reply:
I've implemented the logging tests.